apis/console/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+#![no_std]
+
+use core::fmt;
+use core::marker::PhantomData;
+use libtock_platform::allow_ro;
+use libtock_platform::allow_ro::AllowRo;
+use libtock_platform::share;
+use libtock_platform::subscribe::Subscribe;
+use libtock_platform::{ErrorCode, Syscalls};
+
+/// The console driver.
+///
+/// It allows libraries to pass strings to the kernel's console driver.
+///
+/// # Example
+/// ```ignore
+/// use libtock2::Console;
+///
+/// // Prints 0x45 and the app which called it.
+/// Console::print(0x45);
+/// ```
+pub struct Console<S: Syscalls>(S);
+
+impl<S: Syscalls> Console<S> {
+    /// Run a check against the low-level debug capsule to ensure it is present.
+    ///
+    /// Returns `true` if the driver was present. This does not necessarily mean
+    /// that the driver is working, as it may still fail to allocate grant
+    /// memory.
+    #[inline(always)]
+    pub fn driver_check() -> bool {
+        S::command(DRIVER_NUM, command::DRIVER_CHECK, 0, 0).is_success()
+    }
+
+    /// Writes bytes, returns count of bytes written.
+    pub fn write(s: &[u8]) -> Result<u32, ErrorCode> {
+        let called = core::cell::Cell::new(Option::<(u32,)>::None);
+        share::scope::<(AllowRo<_, DRIVER_NUM, 1>, Subscribe<_, DRIVER_NUM, 1>), _, _>(|handle| {
+            let (allow_ro, subscribe) = handle.split();
+
+            S::allow_ro::<AllowConfig, DRIVER_NUM, 1>(allow_ro, s)?;
+
+            S::subscribe::<
+                _,
+                _,
+                subscribe::Config,
+                DRIVER_NUM,
+                1, //subscribe::WRITE, // this confuses the compiler
+            >(subscribe, &called)?;
+
+            S::command(DRIVER_NUM, command::WRITE, s.len() as u32, 0).to_result()?;
+
+            loop {
+                S::yield_wait();
+                if let Some((bytes_read_count,)) = called.get() {
+                    return Ok(bytes_read_count);
+                }
+            }
+        })
+    }
+
+    /// Writes all bytes of a slice.
+    /// This is an alternative to `fmt::Write::write`
+    /// becaus this can actually return an error code.
+    /// It's makes only one `subscribe` call,
+    /// as opposed to calling `write` in a loop.
+    fn write_all(s: &[u8]) -> Result<(), ErrorCode> {
+        let called = core::cell::Cell::new(Option::<(u32,)>::None);
+        share::scope::<(AllowRo<_, DRIVER_NUM, 1>, Subscribe<_, DRIVER_NUM, 1>), _, _>(|handle| {
+            let (allow_ro, subscribe) = handle.split();
+
+            S::subscribe::<
+                _,
+                _,
+                subscribe::Config,
+                DRIVER_NUM,
+                1, //subscribe::WRITE, // this confuses the compiler
+            >(subscribe, &called)?;
+
+            let mut remaining = s.len();
+            while remaining > 0 {
```

Comment:
Is it possible for this to loop more than once? It looks like the Console syscall driver handles the retry loop for us.

Reply:
I'm not seeing anything in the API that prevents an implementation of the driver to write less than requested.

Reply:
My response was going to be "I looked at the implementation and I've sent a PR to change the documentation and guarantee it always writes the full data requested". However, when I looked at the implementation I had more questions, so I've opened up a thread on Slack to understand what exactly the console capsule is supposed to do.

Regardless, we should update the console syscall driver documentation to remove this ambiguity, then we can update this PR to match.

Reply:
Now that https://github.com/tock/tock/pull/2996 is merged, would you mind removing `write_all`?

Reply:
Done.